segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -521,66 +577,68 @@ private void persistMetadata(LogMetadata metadata, boolean create) throws Durabl
      * 5. Close the previous ledger (and implicitly seal it).
      * 5.1 If closing fails, there is nothing we can do. We've already opened a new ledger and new writes are going to it.
+     *
+     * NOTE: this method is not thread safe and is not meant to be executed concurrently. It should only be invoked as
+     * part of the Rollover Processor.
      */
-    private void rollover() throws DurableDataLogException {
+    private void rollover() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, this.traceObjectId, "rollover");
+        val l = getWriteLedger().ledger;
+        if (l.getLength() < this.config.getBkLedgerMaxSize()) {
+            LoggerHelpers.traceLeave(log, this.traceObjectId, "rollover", traceId, false);
+            return;
+        }
 
-        // Create new ledger.
-        LedgerHandle newLedger = Ledgers.create(this.bookKeeper, this.config);
-        log.debug("{}: Rollover: created new ledger {}.", this.traceObjectId, newLedger.getId());
+        try {
+            // Create new ledger.
+            LedgerHandle newLedger = Ledgers.create(this.bookKeeper, this.config);
+            log.debug("{}: Rollover: created new ledger {}.", this.traceObjectId, newLedger.getId());
 
-        // Update the metadata.
-        LogMetadata metadata = getLogMetadata();
-        metadata = updateMetadata(metadata, newLedger);
-        LedgerMetadata ledgerMetadata = metadata.getLedger(newLedger.getId());
-        assert ledgerMetadata != null : "cannot find newly added ledger metadata";
-        log.debug("{}: Rollover: updated metadata '{}.", this.traceObjectId, metadata);
+            // Update the metadata.
+            LogMetadata metadata = getLogMetadata();
+            metadata = updateMetadata(metadata, newLedger);
+            LedgerMetadata ledgerMetadata = metadata.getLedger(newLedger.getId());
+            assert ledgerMetadata != null : "cannot find newly added ledger metadata";
+            log.debug("{}: Rollover: updated metadata '{}.", this.traceObjectId, metadata);
 
-        // Update pointers to the new ledger and metadata.
-        LedgerHandle oldLedger;
-        synchronized (this.lock) {
-            oldLedger = this.writeLedger.ledger;
-            this.writeLedger = new WriteLedger(newLedger, ledgerMetadata);
-            this.logMetadata = metadata;
+            // Update pointers to the new ledger and metadata.
+            LedgerHandle oldLedger;
+            synchronized (this.lock) {
+                oldLedger = this.writeLedger.ledger;
+                this.writeLedger = new WriteLedger(newLedger, ledgerMetadata);
+                this.logMetadata = metadata;
+            }
+
+            // Close the old ledger. This must be done outside of the lock, otherwise the pending writes (and their callbacks
+            // will be invoked within the lock, thus likely candidates for deadlocks).
+            Ledgers.close(oldLedger);
```

Comment:
I'm worried about this step here. We must be absolutely sure of what goes into the ledger we are closing. Specifically, say that we close the old ledger while writing to the ledger concurrently. There are two possible outcomes:
- It closes correctly and we can determine what has gone in by calling `getLastAddConfirmed`. That will give us the number of the last entry that has gone in.
- The close call fails say because of a network partition. In this case,  another bk client could end up recovering the ledger and we won't know precisely what happened to the ledger until we are able to open it and likewise call `getLasAddConfirmed`.

In the heuristic proposed here:

http://bookkeeper.apache.org/docs/r4.4.0/bookkeeperLedgers2Logs.html

The step 4 of rollover isn't very specific about what happens in the case the close fails. The main issue we have is that if the bk client is unable to close the ledger cleanly, then it is possible that some other client will open and recover it, in which case the ledger is brought to a consistent state, but we don't really know what that state is unless we read it. In bk-land, we often use the entry ids to track what has gone in.

Note that even if we only perform the close once we have flushed all pending adds, we can still be left in an uncertain state with respect to the last write. Given that such scenarios should be rare, the right thing to do seems to be perform recovery over the tier one container data. We declare the failure to close cleanly a ledger during rollover as a fatal one and perform recovery.